dask_expr/_core.py

```diff
@@ -30,23 +35,36 @@ class Expr:
     _defaults = {}
     _instances = weakref.WeakValueDictionary()
 
-    def __new__(cls, *args, **kwargs):
+    def __new__(cls, *args, _branch_id=None, **kwargs):
         operands = list(args)
+        if _branch_id is None and len(operands) and isinstance(operands[-1], BranchId):
+            _branch_id = operands.pop(-1)
+        elif _branch_id is None:
+            _branch_id = BranchId(0)
+
         for parameter in cls._parameters[len(operands) :]:
             try:
                 operands.append(kwargs.pop(parameter))
             except KeyError:
                 operands.append(cls._defaults[parameter])
         assert not kwargs, kwargs
         inst = object.__new__(cls)
-        inst.operands = [_unpack_collections(o) for o in operands]
+        inst.operands = [_unpack_collections(o) for o in operands] + [_branch_id]
         _name = inst._name
         if _name in Expr._instances:
             return Expr._instances[_name]
 
         Expr._instances[_name] = inst
         return inst
 
+    @functools.cached_property
+    def argument_operands(self):
+        return self.operands[:-1]
+
+    @functools.cached_property
+    def _branch_id(self):
+        return self.operands[-1]
```

Comment:
How about just storing this as an attribute? Why does this need to be an operand?

Reply:
Having it in the operands makes sure that it is used for ``_name``, which is what we mostly care about. If that is overridden and forgotten, then the whole thing won't work properly

Reply:
I get that. I'm just concerned this is messing with other things since `operands` are everywhere. If it works I'm fine.

Reply:
I think to "avoid" the kind of confusion I'm thinking about you introduced argument_opearands but this may also be confusing to developers